src/app/collection/settings/tab/rules/add-rule-form/blockly-form/blockly-editor/blockly-editor.component.ts

```diff
@@ -0,0 +1,792 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import {
+  AfterViewChecked,
+  AfterViewInit,
+  ChangeDetectionStrategy,
+  Component,
+  ElementRef,
+  EventEmitter,
+  Inject,
+  Input,
+  OnDestroy,
+  Output,
+  Renderer2,
+  ViewChild,
+} from '@angular/core';
+import {environment} from '../../../../../../../../environments/environment';
+import {Store} from '@ngrx/store';
+import {AppState} from '../../../../../../../core/store/app.state';
+import {ActivatedRoute} from '@angular/router';
+import {DialogService} from '../../../../../../../dialog/dialog.service';
+import {DOCUMENT} from '@angular/common';
+import {LinkType} from '../../../../../../../core/store/link-types/link.type';
+import {Collection} from '../../../../../../../core/store/collections/collection';
+import {Variable} from '../../../variable-type';
+import {shadeColor} from '../../../../../../../shared/utils/html-modifier';
+import {COLOR_DARK, COLOR_GRAY200, COLOR_GREEN, COLOR_PRIMARY, COLOR_RED} from '../../../../../../../core/constants';
+import {ContrastColorPipe} from '../../../../../../../shared/pipes/contrast-color.pipe';
+
+declare var Blockly: any;
+
+@Component({
+  selector: 'blockly-editor',
+  templateUrl: './blockly-editor.component.html',
+  styleUrls: ['./blockly-editor.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class BlocklyEditorComponent implements AfterViewInit, OnDestroy {
+  @Input('collections')
+  public collections: Collection[] = [];
+
+  @Input('linkTypes')
+  public linkTypes: LinkType[] = [];
+
+  @Input('variables')
+  public variables: Variable[] = [];
+
+  @Input('xml')
+  public xml: string = '';
+
+  @ViewChild('loading')
+  private loadingElement: ElementRef;
+
+  @Output()
+  public onJsUpdate = new EventEmitter<string>();
+
+  @Output()
+  public onXmlUpdate = new EventEmitter<string>();
+
+  public blocklyId = String(Math.floor(Math.random() * 1000000000000000) + 1);
+
+  private workspace: any;
+  private lumeerVar: string;
+
+  public static THESE = new Map();
+
+  private static readonly DOCUMENT_TYPE_SUFFIX = '_document';
+  private static readonly DOCUMENT_ARRAY_TYPE_SUFFIX = '_document_array';
+  private static readonly LINK_TYPE_SUFFIX = '_link';
+  private static readonly ARRAY_TYPE_SUFFIX = '_array';
+  private static readonly FOREACH_DOCUMENT_ARRAY = 'foreach_document_array';
+  private static readonly GET_ATTRIBUTE = 'get_attribute';
+  private static readonly SET_ATTRIBUTE = 'set_attribute';
+  private static readonly VARIABLES_GET_PREFIX = 'variables_get_';
+  private static readonly UNKNOWN = 'unknown';
+  private static readonly STATEMENT_CONTAINER = 'statement_container';
```

Comment:
You can define these strings as constants out of class and you then can rid of preffix `BlocklyEditorCompont` in every usage.

```
const LINK_TYPE_SUFFIX = '_link'; (without export!)
.....
.....
export class BlocklyEditorComponent {
.....
}
````

Reply:
I am probably fine with the local namespace...

Reply:
But it is used many times so I recommend you to do it.